File: services/torchserveClient.js
/**
 * TorchServe Client - Handles communication with external prediction API
 */

const fetch = require('node-fetch');

// Import URLs from schema
const PREDICTION_URL = 'https://anomalib-serving-360893797389.europe-central2.run.app/predictions/model';
const HEALTH_URL = 'https://anomalib-serving-360893797389.europe-central2.run.app/ping';

class TorchServeClient {
  constructor(baseUrl) {
    // Use external API URLs from schema.js instead of local TorchServe
    this.baseUrl = baseUrl || PREDICTION_URL.replace('/predictions/model', '');
    this.predictUrl = PREDICTION_URL;
    this.healthUrl = HEALTH_URL;
  }

  /**
   * Check if prediction service is healthy using /ping endpoint
   */
  async checkHealth() {
    try {
      const response = await fetch(this.healthUrl, {
        method: 'GET',
        timeout: 5000
      });

      if (response.ok) {
        const text = await response.text();
        // Check if response contains expected health indicator
        return {
          healthy: true,
          status: response.status,
          message: text
        };
      } else {
        return {
          healthy: false,
          status: response.status,
          error: `HTTP ${response.status}`
        };
      }
    } catch (error) {
      return {
        healthy: false,
        error: error.message
      };
    }
  }

  /**
   * Predict anomaly from image buffer
   * @param {Buffer} imageBuffer - JPEG/PNG image buffer
   * @param {Object} options - Prediction options
   * @param {number} options.threshold - Anomaly threshold (0-1)
   * @param {boolean} options.includeOverlay - Include visualization overlay
   * @returns {Promise<Object>} Prediction result
   */
  async predict(imageBuffer, options = {}) {
    const {
      threshold = 0.5,
      includeOverlay = false
    } = options;

    try {
      // Validate image buffer
      if (!imageBuffer || !Buffer.isBuffer(imageBuffer)) {
        console.error('[TorchServe] Validation failed: Invalid image buffer type');
        throw new Error('Invalid image buffer provided');
      }

      if (imageBuffer.length === 0) {
        console.error('[TorchServe] Validation failed: Empty image buffer');
        throw new Error('Image buffer is empty');
      }

      if (imageBuffer.length < 100) { // Minimum valid image size
        console.error('[TorchServe] Validation failed: Image buffer too small (likely corrupted)');
        throw new Error('Image buffer too small - likely corrupted or incomplete');
      }

      if (imageBuffer.length > 50 * 1024 * 1024) { // 50MB limit
        console.error(`[TorchServe] Validation failed: Image too large (${imageBuffer.length} bytes)`);
        throw new Error(`Image too large: ${(imageBuffer.length / 1024 / 1024).toFixed(2)}MB (max 50MB)`);
      }

      // Validate image format by checking magic bytes
      const isValidImage = this.validateImageFormat(imageBuffer);
      if (!isValidImage) {
        console.error('[TorchServe] Validation failed: Invalid image format (not JPEG/PNG)');
        throw new Error('Invalid image format - must be JPEG or PNG');
      }

      // Validate threshold
      if (typeof threshold !== 'number' || threshold < 0 || threshold > 1) {
        console.error(`[TorchServe] Validation failed: Invalid threshold value: ${threshold}`);
        throw new Error(`Invalid threshold: ${threshold} (must be 0-1)`);
      }

      // Validate options types
      if (typeof includeOverlay !== 'boolean') {
        console.error('[TorchServe] Validation failed: includeOverlay must be boolean');
        throw new Error('includeOverlay must be a boolean value');
      }

      // Convert image buffer to base64
      let imageBase64;
      try {
        imageBase64 = imageBuffer.toString('base64');
      } catch (err) {
        console.error('[TorchServe] Failed to encode image to base64:', err);
        throw new Error('Failed to encode image data');
      }

      // Validate base64 encoding
      if (!imageBase64 || imageBase64.length === 0) {
        console.error('[TorchServe] Validation failed: Base64 encoding resulted in empty string');
        throw new Error('Failed to encode image - empty result');
      }

      // Prepare payload matching your handler's expected format
      const payload = {
        data: imageBase64,
        threshold: threshold,
        include_overlay: includeOverlay
      };

      // Validate payload structure before sending
      if (!payload.data || typeof payload.data !== 'string') {
        console.error('[TorchServe] Validation failed: Invalid payload structure');
        throw new Error('Invalid payload structure - missing or invalid data field');
      }

      const payloadSize = JSON.stringify(payload).length;
      console.log(`[TorchServe] Sending prediction request:`);
      console.log(`  - Image size: ${(imageBuffer.length / 1024).toFixed(2)}KB`);
      console.log(`  - Payload size: ${(payloadSize / 1024).toFixed(2)}KB`);
      console.log(`  - Threshold: ${threshold}`);
      console.log(`  - Include overlay: ${includeOverlay}`);
      console.log(`  - URL: ${this.predictUrl}`);

      const startTime = Date.now();

      // Send POST request to TorchServe
      const response = await fetch(this.predictUrl, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify(payload),
        timeout: 60000 // 60 second timeout for inference (increased from 30s)
      });

      const requestTime = Date.now() - startTime;
      console.log(`[TorchServe] Request completed in ${requestTime}ms (status: ${response.status})`);

      if (!response.ok) {
        const errorText = await response.text();
        console.error(`[TorchServe] Error response:`, errorText);
        throw new Error(`TorchServe error (${response.status}): ${errorText}`);
      }

      const result = await response.json();
      console.log(`[TorchServe] Prediction received:`, JSON.stringify(result, null, 2));

      // Parse TorchServe response format
      // Your handler returns: { predictions: [...], predicted_class: "...", threshold: 0.5, overlay: "..." }
      return this.parseResponse(result);

    } catch (error) {
      if (error.type === 'request-timeout') {
        console.error(`[TorchServe] Request timed out after ${error.timeout || 60000}ms`);
        throw new Error(`Prediction service timed out. The service may be slow or unavailable.`);
      }
      console.error(`[TorchServe] Prediction failed:`, error);
      throw new Error(`Prediction failed: ${error.message}`);
    }
  }

  /**
   * Validate image format by checking magic bytes
   * @param {Buffer} buffer - Image buffer to validate
   * @returns {boolean} True if valid JPEG or PNG
   */
  validateImageFormat(buffer) {
    // Check JPEG magic bytes (FF D8 FF)
    if (buffer[0] === 0xFF && buffer[1] === 0xD8 && buffer[2] === 0xFF) {
      return true;
    }

    // Check PNG magic bytes (89 50 4E 47 0D 0A 1A 0A)
    if (buffer[0] === 0x89 && buffer[1] === 0x50 && buffer[2] === 0x4E && buffer[3] === 0x47) {
      return true;
    }

    return false;
  }

  /**
   * Parse TorchServe response to standardized format
   */
  parseResponse(torchserveResult) {
    // Handle array response (TorchServe may wrap in array)
    const result = Array.isArray(torchserveResult) ? torchserveResult[0] : torchserveResult;

    // Extract anomaly score from predictions array
    const anomalyPrediction = result.predictions?.find(p => p.class === 'anomalous');
    const anomalyScore = anomalyPrediction?.probability || 0;

    // Return standardized format matching what UI expects
    return {
      anomaly_score: anomalyScore,
      is_anomaly: result.predicted_class === 'anomalous',
      predicted_class: result.predicted_class,
      threshold_used: result.threshold,
      predictions: result.predictions,
      overlay: result.overlay || null,
      inference_time_ms: 0  // TorchServe doesn't provide this, could add timing here
    };
  }

  /**
   * Get model information
   */
  async getModelInfo() {
    try {
      const response = await fetch(`${this.baseUrl}/models/model`, {
        timeout: 5000
      });

      if (response.ok) {
        return await response.json();
      }

      return null;
    } catch (error) {
      // Silently fail - connection errors are expected when TorchServe is not running locally
      return null;
    }
  }
}

module.exports = TorchServeClient;

File: services/torchserve_client.py
"""
TorchServe Client - Handles communication with Anomalib TorchServe container
"""

import base64
import requests
import json
import io
# from PIL import Image  # Commented out as preprocess_image is commented
# import torchvision.transforms as transforms  # Commented out as preprocess_image is commented
# import torchvision.transforms.functional as F  # Commented out as preprocess_image is commented

# # ImageNet normalization parameters  # Commented out as preprocess_image is commented
# IMAGENET_MEAN = [0.485, 0.456, 0.406]  # Commented out as preprocess_image is commented
# IMAGENET_STD = [0.229, 0.224, 0.225]  # Commented out as preprocess_image is commented


class TorchServeClient:
    def __init__(self, base_url=None):
        self.base_url = base_url or 'http://prediction-server:8080'
        self.predict_url = f'{self.base_url}/predictions/model'
        self.health_url = f'{self.base_url}/ping'

    # def preprocess_image(self, image_buffer):
    #     """
    #     Preprocess image for Patchcore inference.
    #     Resizes to 256x256, normalizes with ImageNet parameters, and converts to PNG.

    #     Args:
    #         image_buffer: bytes - Raw image data (JPEG/PNG)

    #     Returns:
    #         str - Base64 encoded preprocessed image
    #     """
    #     # Open and convert image to RGB
    #     pil_image = Image.open(io.BytesIO(image_buffer)).convert('RGB')

    #     # Create preprocessing transform
    #     transform = transforms.Compose([
    #         transforms.Resize((256, 256), interpolation=transforms.InterpolationMode.BICUBIC),
    #         transforms.ToTensor(),
    #         transforms.Normalize(mean=IMAGENET_MEAN, std=IMAGENET_STD)
    #     ])

    #     # Apply transform
    #     processed_tensor = transform(pil_image)

    #     # Convert tensor back to PIL image and then to base64 PNG
    #     buffer = io.BytesIO()
    #     F.to_pil_image(processed_tensor).save(buffer, format='PNG')
    #     base64_image = base64.b64encode(buffer.getvalue()).decode('utf-8')

    #     return base64_image

    def check_health(self):
        """Check if TorchServe is healthy"""
        try:
            response = requests.get(self.health_url, timeout=5)
            return {
                'healthy': response.ok,
                'status': response.status_code
            }
        except Exception as error:
            return {
                'healthy': False,
                'error': str(error)
            }

    def predict(self, image_buffer, options=None):
        """
        Predict anomaly from image buffer

        Args:
            image_buffer: bytes - JPEG/PNG image buffer
            options: dict - Prediction options
                - threshold: float - Anomaly threshold (0-1)
                - include_overlay: bool - Include visualization overlay

        Returns:
            dict - Prediction result
        """
        if options is None:
            options = {}

        threshold = options.get('threshold', 0.5)
        include_overlay = options.get('include_overlay', False)

        try:
            # Directly convert image to base64 without preprocessing
            print(f'[TorchServe] Encoding image...')
            image_base64 = base64.b64encode(image_buffer).decode('utf-8')
            print(f'[TorchServe] Image encoded (base64 length: {len(image_base64)} chars)')

            # Prepare payload matching handler's expected format
            payload = {
                'data': image_base64,
                'threshold': threshold,
                'include_overlay': include_overlay,
                'image_size': None  # No preprocessing, original image size
            }

            print(f'[TorchServe] Sending prediction request to {self.predict_url}')
            print(f'[TorchServe] Parameters: threshold={threshold}, overlay={include_overlay}')

            # Send POST request to TorchServe (increased timeout for remote servers)
            response = requests.post(
                self.predict_url,
                headers={'Content-Type': 'application/json'},
                json=payload,  # Use json parameter instead of data + json.dumps
                timeout=120  # Increased to 2 minutes for remote servers
            )

            print(f'[TorchServe] Response status: {response.status_code}')

            if not response.ok:
                error_text = response.text
                raise Exception(f'TorchServe error ({response.status_code}): {error_text}')

            result = response.json()
            print(f'[TorchServe] Prediction received: {json.dumps(result, indent=2)}')

            # Parse TorchServe response format
            return self.parse_response(result)

        except requests.exceptions.ConnectionError as error:
            print(f'[TorchServe] Connection failed to {self.predict_url}: {error}')
            raise Exception(f'Cannot connect to TorchServe at {self.predict_url}. Is the prediction service running?')
        except requests.exceptions.Timeout as error:
            print(f'[TorchServe] Request timeout: {error}')
            raise Exception('Prediction request timed out after 2 minutes. The remote server may be cold-starting or overloaded.')
        except Exception as error:
            print(f'[TorchServe] Prediction failed: {error}')
            raise Exception(f'Prediction failed: {error}')

    def parse_response(self, torchserve_result):
        """Parse TorchServe response to standardized format"""
        # Handle array response (TorchServe may wrap in array)
        result = torchserve_result[0] if isinstance(torchserve_result, list) else torchserve_result

        # Extract anomaly score from predictions array
        anomaly_prediction = None
        if result.get('predictions'):
            anomaly_prediction = next(
                (p for p in result['predictions'] if p.get('class') == 'anomalous'),
                None
            )
        anomaly_score = anomaly_prediction.get('probability', 0) if anomaly_prediction else 0

        # Return standardized format matching what UI expects
        return {
            'anomaly_score': anomaly_score,
            'is_anomaly': result.get('predicted_class') == 'anomalous',
            'predicted_class': result.get('predicted_class'),
            'threshold_used': result.get('threshold'),
            'predictions': result.get('predictions'),
            'overlay': result.get('overlay'),
            'inference_time_ms': 0  # TorchServe doesn't provide this
        }

